src/main/scala/pbdirect/PBReader.scala

```diff
@@ -53,14 +77,24 @@ trait LowerPriorityPBReaderImplicits {
     gen.from(repr.value.parse(1, out.toByteArray))
   }
 }
+
 trait PBReaderImplicits extends LowerPriorityPBReaderImplicits {
 
-  implicit def prodReader[A, R <: HList](
+  object collectFieldIndices extends Poly1 {
+    implicit val defaultCase = at[Some[pbIndex]] {
+      case Some(annotation) => FieldIndex(annotation.first :: annotation.more.toList)
+    }
+  }
+
```

Comment:
This is similar to the writer, but in reverse.

We use the annotations to build an `HList` of `FieldIndex`es:

```scala
FieldIndex(1) :: FieldIndex(2) :: HNil
```

and give that `HList` to the `PBProductReader`, along with the input bytes.

`consProductReader` makes use of the head of the `HList` of `FieldIndex`es to see what index it should be looking for, then passes the tail of the `HList` in the recursive step.